text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,590 @@
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more messages are passed to a consumer. |
+| `settle`       | One or more messages are settled. |
+
+For further details about each of those operations refer to the [section about trace structure](#trace-structure).
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+
+The kind of `settle` spans should be set according to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
+e. g. it should be to `CLIENT` if the `settle` spans models a synchronous call
+to the intermediary.
+
+Setting span kinds according to this table ensures, that span links between
+consumers and producers always go from a `PRODUCER` span on the producer side to
+a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
+linked traces without the need for additional semantic hints.
+
+### Span relationships
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
+SHOULD account only for a single message.
+
+If a "Create" span exists for a message, its context SHOULD be attached to
+the message. If no "Create" span exists, the context of the related "Publish"
+span SHOULD be attached to the message.
+
+#### Consumer spans
+
+##### Push-based scenarios
+
+"Deliver" spans SHOULD be created for operations of passing messages to the
+application when those operations are not initiated by the application
+code.
+
+##### Pull-based scenarios
+
+"Receive" spans SHOULD be created for operations of passing messages to the
+application when those operations are initiated by the application code.
+
+##### General considerations
+
+"Deliver" or "Receive" spans MUST NOT be created for messages which are not
+forwarded to the caller, but are pre-fetched or cached by messaging
+libraries or SDKs.
+
+A single "Deliver" or "Receive" span can account for a single message, for
+multiple messages (in case messages are passed for processing as batches), or
+for no message at all (if it is signalled that no messages were received).  For
+each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
+the "Create" or "Publish" span for the message. In addition, if it is possible
```

Comment:
If we use the "Publish" span as the "create" context of all messages, then here we will be essentially adding duplicated links, right? Is this something we discussed at some point?

Reply:
In the context of publishing batches, this will result in having the same context for different messages. I think that's what you mean.

We discussed this, but I can make it more explicit in the text if needed.

Reply:
> In the context of publishing batches, this will result in having the same context for different messages. I think that's what you mean.

Exactly. If you read it in full it becomes clear. My point was more regarding adding links that all point to the same thing. But I guess there's no easy way around that. We can discuss during the meeting, but I fear there's no way around it.